custom_components/hubitat/fan.py

```diff
@@ -0,0 +1,85 @@
+"""Support for Hubitat fans."""
+
+from logging import getLogger
+import re
+from typing import Any, List, Optional
+
+from hubitatmaker import (
+    ATTR_SPEED,
+    CAP_FAN_CONTROL,
+    CMD_SET_LEVEL,
+    CMD_OFF, 
+    CMD_ON, 
+    CMD_CYCLE_SPEED, 
+    CMD_SET_SPEED,
+    Device
+)
+
+from homeassistant.components.fan import (
+    FanEntity
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.util import color as color_util
+
+from .device import Hub, HubitatEntity, get_hub
+_LOGGER = getLogger(__name__)
+
+class HubitatFan(HubitatEntity, FanEntity):
+    """Representation of a Hubitat fan."""
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if the switch is on."""
+        return self.get_str_attr("switch") == "on"
+
+    # @property
+    # def brightness(self) -> Optional[int]:
+    #     """Return the level of this fan."""
+    #     level = self.get_int_attr("level")
+    #     if level is None:
+    #         return None
+    #     return int(255 * level / 100)
+
+    @property
+    def speed(self) -> str:
+        """Return the speed of this fan."""
+        return self.get_str_attr(ATTR_SPEED)
+
+    @property
+    def speed_list(self) -> str:
+        """Return the list of speeds for this fan. (MakerAPI isn't sending the list of speeds)"""
+        return ["off", "low", "medium-low", "medium", "high", "auto"]
```

Comment:
At least the virtual fan controller does return a list of values in a `values` property on the `speed` attribute. 

```json
{
   "id" : "2087",
   "attributes" : [
      {
         "name" : "speed",
         "dataType" : "ENUM",
         "values" : [
            "low",
            "medium-low",
            "medium",
            "medium-high",
            "high",
            "on",
            "off",
            "auto"
         ],
         "currentValue" : "off"
      }
   ],
   "name" : "Virtual Fan Controller",
 ```

In an ideal world, we'd try to use the list from the device first, and fall back to a hard coded list if necessary. 
You'd need to add a `values` property to the Attribute class in hubitatmaker for that to work. Then you'd end up with something like:

```python
return self._device.attributes[ATTR_SPEED].values or DEFAULT_SPEEDS
```

Reply:
I have a few of the Hampton Bay fan controllers, and the MakerAPI doesn't return the speed values.  For some reason, they return the switch values (on/off) instead:
```
{
    "name": "Hampton Bay Fan Component",
    "label": "Bedroom Fan",
    "type": "Hampton Bay Fan Component",
    "id": "117",
    "date": "2020-02-25T02:12:10+0000",
    "model": null,
    "manufacturer": null,
    "capabilities": [
        "Switch",
        "SwitchLevel",
        "Refresh",
        "Actuator",
        "FanControl"
    ],
    "attributes": {
        "level": "50",
        "dataType": "ENUM",
        "values": [
            "on",
            "off"
        ],
        "speed": "medium-low",
        "switch": "on"
    },
    "commands": [
        {
            "command": "cycleSpeed"
        },
        {
            "command": "off"
        },
        {
            "command": "on"
        },
        {
            "command": "refresh"
        },
        {
            "command": "setLevel"
        },
        {
            "command": "setSpeed"
        }
    ]
}
```

Reply:
Wait... that's in the "Get All Devices with Full Details" link

Reply:
Getting just the device info returns this:
```
{
    "id": "140",
    "name": "Hampton Bay Fan Component",
    "label": "Office Fan",
    "attributes": [
        {
            "name": "level",
            "currentValue": 25,
            "dataType": "NUMBER"
        },
        {
            "name": "speed",
            "currentValue": "off",
            "dataType": "ENUM",
            "values": [
                "low",
                "medium-low",
                "medium",
                "medium-high",
                "high",
                "on",
                "off",
                "auto"
            ]
        },
        {
            "name": "switch",
            "currentValue": "off",
            "dataType": "ENUM",
            "values": [
                "on",
                "off"
            ]
        }
    ],
    "capabilities": [
        "Switch",
        {
            "attributes": [
                {
                    "name": "switch",
                    "dataType": null
                }
            ]
        },
        "SwitchLevel",
        {
            "attributes": [
                {
                    "name": "level",
                    "dataType": null
                }
            ]
        },
        "Refresh",
        "Actuator",
        "FanControl",
        {
            "attributes": [
                {
                    "name": "speed",
                    "dataType": null
                }
            ]
        }
    ],
    "commands": [
        "cycleSpeed",
        "off",
        "on",
        "refresh",
        "setLevel",
        "setSpeed"
    ]
}
```

Reply:
Would this work for the `values` property?
```
    @property
    def values(self) -> Tuple[str, ...]:
        return tuple(self._properties["values"])
```

Reply:
Or, would this be better?
```
    @property
    def values(self) -> Tuple[str, ...]:
        return tuple(self._properties.get("values", []))
```

Reply:
I was basing it off of the code in `update_state` that handles capabilities.

Reply:
Maybe do this, since `values` is an optional property, and it may be useful to differentiate between an empty values list and a missing values attribute.

```python
@property
def values(self) -> Optional[Tuple[str, ...]]:
  if "values" not in self._properties:
    return None
  return tuple(self._properties["values"])
```

For just starting out in Python, you've picked up the syntax and typing system pretty quickly. 😄

Reply:
Thanks!  Those changes should be in there now.